torch/csrc/jit/pybind_utils.h

```diff
@@ -77,6 +78,69 @@ inline TypedIValue toDictKeyIValue(py::handle key) {
   }
 }
 
+inline bool isNamedTupleClass(py::object obj) {
+  auto tuple_type = reinterpret_cast<PyObject*>(&PyTuple_Type);
+  return PyObject_IsSubclass(obj.ptr(), tuple_type) &&
+      py::hasattr(obj, "_fields");
+}
+
+inline TypePtr getOrCreateNamedTupleType(
+  py::object obj,
+  const c10::QualifiedName& qualName,
+  c10::optional<std::vector<TypePtr>> element_types,
+  const SourceRange& loc) {
+  // Currently don't support default values
+  if (py::hasattr(obj, "_field_defaults")) {
+    auto default_dict = py::cast<std::map<std::string, py::object>>(
+        py::getattr(obj, "_field_defaults"));
+    if (default_dict.size()) {
+      std::string error_msg =
+          "Default values are currently not supported"
+          " on NamedTuple fields in TorchScript. Fields "
+          "with default values: [";
+      bool first = true;
+      for (const auto& kv : default_dict) {
+        if (!first) {
+          error_msg += ", ";
+        }
+        error_msg += kv.first;
+      }
+      error_msg += "]";
+      throw script::ErrorReport(loc) << error_msg;
+    }
+  }
+
+  std::string unqualName;
+  std::vector<std::string> fields;
+  c10::NamedTypePtr named_type;
+  py::object props;
+  if (element_types.has_value()) {
+    props = py::module::import("torch.jit")
+                .attr("_get_named_tuple_properties")(obj, /*need_types=*/false);
+    std::tie(unqualName, fields) = py::cast<
+        std::tuple<std::string, decltype(fields)>>(props);
+    named_type = TupleType::createNamed(qualName, fields, element_types.value());
+  } else {
+    std::vector<TypePtr> annotations;
+    props = py::module::import("torch.jit")
+                .attr("_get_named_tuple_properties")(obj, /*need_types=*/true);
+    std::tie(unqualName, fields, annotations) = py::cast<
+        std::tuple<std::string, decltype(fields), decltype(annotations)>>(props);
+    named_type = TupleType::createNamed(qualName, fields, annotations);
+  }
+
+  if (auto type = get_python_cu()->get_type(qualName)) {
```

Comment:
This is copied from init.cpp, where I think the logic is if we use the same named tuple in the trace:
* first time it will create the named tuple type (`named_type`) as normal.
* second time we hit the same named tuple type, we first still create the `named_type`, then we check if the python_cu already contains `type` that is subtype of `named_type`, if we do, then we return `type`, and drop the newly created `named_type`

Reply:
It looks like @suo added this particular check: https://github.com/pytorch/pytorch/pull/23798

NamedTuple types are unique'd via QualifiedName. Does this answer your question @zdevito